src/cache.rs

```diff
@@ -226,33 +307,104 @@ impl<T: Hash + PartialEq> Constraint<T> {
     /// Enter a constraint for a call to an immutable function.
```

Comment:
The comment here is mistleading. Not your fault, but maybe you could still update it.